### ISSUE TITLE
Preparing for Aztlan hard fork issues

### ISSUE THREAD

**Issue by sorpaas** [ethereumclassic/ECIPs#217]
Recently, a few community members have decided to implement the Aztlan hard fork. The issue with Aztlan is that without EIP-1884, it effectively left an account trie attack vector open. Most current clients do not have other protections currently other than increasing gas cost. The paper for the attack is actually public, meaning anyone who knows the method can do it. As the security analysis for Aztlan is inadequate, we should prepare for its issues properly.

My recommendations would be we do not implement Aztlan before account versioning. However, given the consensus was already reached, I won't oppose the hard fork.

**Comment by bobsummerwill** [comment]
You realize that the outcome here, @sorpaas, is going to be the effective death of Multi-Geth?

Classic Geth will be END-of-LIFE after Agharta.
You can obviously add support to Parity-Ethereum.
Hyperledger Besu won't support your "Ethereum Classic Classic" flavor.

So Ethereum Classic will end up with Parity-Ethereum (majority client) and Hyperledger Besu (minority client) only.    Maybe I can get Mantis revived too by March as a third client.

But Multi-Geth will only be used on Ethereum Classic Classic.

I actually personally like that outcome for clients, to be honest, because Geth has an unsolvable backlog of IP risks (no known provenance for contributions, GPLv3 licensing which restricts who can use it).

I have long stated that I think the whole Geth-family will likely die out in the next couple of years because of these IP issues:

- Geth
- Classic Geth
- Multi-Geth
- TurboGeth (will prove @ledgerwatch's ideas on data-structures, but not a long life itself)
- JPM Quorum
- Clearmatics Autonity
- Axioni AxCore

Your actions here would just accelerate that process.

**Comment by bobsummerwill** [comment]
I think there is a decent opportunity here to get TurboGeth as a replacement Geth-family codebase for ETC by March.

https://github.com/ledgerwatch/turbo-geth/issues/162 (by @antsankov)

A community member approached me looking for ways he could help ETC and I pointed him at that potential opportunity.   It would just involve side-porting of ETC support diffs from Multi-Geth into TurboGeth.

First step would be just the ETC support.
Then from there we could do the SHA3/Keccak256 support and Astor testnet support.

@zmitton would you be happy to do SHA3/Keccak256 support against TurboGeth rather than Multi-Geth?  CC @YazzyYaz

**Comment by sorpaas** [comment]
@bobsummerwill I won't mind multi-geth being on Ethereum Classic Classic only at all. After all, I do not have financial involvement in Ethereum Classic, and it's all volunteering work to support a network. If other clients can offload the burden on multi-geth, I would hugely appreciate it.

Besides, if a user do not want to support the contentious hard fork, the only thing they need to do is not to update the client.

Given the nature of contentious hard fork, it's better if you start promoting your fork soon. Otherwise your chain might end up becoming Ethereum Classic Classic.

**Comment by meowsbits** [comment]
> I won't mind multi-geth being on Ethereum Classic Classic only at all

@sorpaas I think the whole point of multi-geth is to be network-extensible, so it should - with adequate and relative developer support - provide configurable use for as many a wide set of EVM-based (and even beyond). _Multi_. I don't care which networks it ends up on; but it should strive to remove opinions for the sake of the code and the technology.

> these IP issues

@bobsummerwill 

I don't see IP issues, I see forks. Varying forks for varying use cases.

**Comment by sorpaas** [comment]
@meowsbits As mentioned, there'll be a flag in multi-geth that supports ECIP-1061. It will just not be Ethereum Classic's default flag.

```
./geth --ecip1061
```

**Comment by bobsummerwill** [comment]
@sorpaas RE: Flag - cool.

./geth --ecip1061

@meowsbits https://bobsummerwill.com/2016/07/12/ethereum-everywhere/

The source code for Geth and all downstreams has unknown provenance.   We do not know the full set of authors and there is zero defence there against future IP claims.   Patent trolling, etc.

GPLv3 repels use by legal entities, which is why there has been a huge shift from copyleft to permissive licensing, from "free software" to "open source" over the last 30 years.

The GPL is still fighting the last war.   It was an awesome tool when it was RMS vs the world.   Now things have done a complete 180 and open source is the default.

Three reasons why IBM created Fabric:

1. Licensing
2. Lack of maturity of Ethereum technology (will it work?)
3. Governance

All three of those had solutions in 2016.   If Gav and Parity had not blocked this relicensing effort then the likely alternative timeline we would have been following would be for IBM to throw away Fabric and to have built IBM Blockchain on top of Ethereum.

IBM had a 30 engineer team ready to pile onto cpp-ethereum when the relicensing was done.   Never happened:

https://www.ibtimes.co.uk/hyperledger-project-reflects-blockchain-politics-1603381

Licensing aside, the bulk of my work those 5 months was chasing all prior contributors and getting consent to the relicencing AND DCO attestions that they had authored the code, and that they had the right to contribute it under that license.   Consent from employers too, if they had them.

Basically getting certainty that no legal claims could be launched in the future.

**Comment by meowsbits** [comment]
> Basically getting certainty that no legal claims could be launched in the future.

Against whom?

> The source code for Geth and all downstreams has unknown provenance. We do not know the full set of authors and there is zero defence there against future IP claims. Patent trolling, etc.

Maybe 11k commits and years worth of available surrounding Github metadata (at least), might suggest a notion a _little less_ boisterous than "unknown provenance" and "zero defense"...  

Are you a lawyer? Have you spoken with one or have documentation from a lawyer about the claims you're making here? Why haven't the patent trolls already taken over? And again, _whose_ interests are we protecting here against these supposed inevitable patent troll lawsuits?

**Comment by meowsbits** [comment]
Oh, and [`3. Governance`](https://github.com/ethereumclassic/ECIPs/issues/217#issuecomment-559881215) -- what was the solution for that which supposedly existed in 2016 for Ethereum? The Foundation?

**Comment by meowsbits** [comment]
Anyways.... @bobsummerwill while an important discussion we're having here -- maybe we're diving down a tangential rabbit hole here. 

Let's try to shift back to the OP's initial topic:

__The issue with Aztlan is that without EIP-1884, it effectively left an account trie attack vector open. Most current clients do not have other protections currently other than increasing gas cost. The paper for the attack is actually public, meaning anyone who knows the method can do it. As the security analysis for Aztlan is inadequate, we should prepare for its issues properly.__

**Comment by sorpaas** [comment]
@meowsbits I created issue https://github.com/ethereumclassic/ECIPs/issues/226 for this purpose!

**Comment by bobsummerwill** [comment]
@meowsbits, I created "Aztlan Fork without EIP-1884 leaves an account trie attack vector open"
https://github.com/ethereumclassic/ECIPs/issues/227

And will continue with the IP chat on #225.
